src/JsonRpcEngine.ts

```diff
@@ -1,69 +1,19 @@
 import SafeEventEmitter from '@metamask/safe-event-emitter';
+import {
+  hasProperty,
+  JsonRpcError,
+  JsonRpcRequest,
+  JsonRpcResponse,
+} from '@metamask/utils';
 import { errorCodes, EthereumRpcError, serializeError } from 'eth-rpc-errors';
 
-type Maybe<T> = Partial<T> | null | undefined;
-
-export type Json =
-  | boolean
-  | number
-  | string
-  | null
-  | { [property: string]: Json }
-  | Json[];
-
-/**
- * A String specifying the version of the JSON-RPC protocol.
- * MUST be exactly "2.0".
- */
-export type JsonRpcVersion = '2.0';
-
-/**
- * An identifier established by the Client that MUST contain a String, Number,
- * or NULL value if included. If it is not included it is assumed to be a
- * notification. The value SHOULD normally not be Null and Numbers SHOULD
- * NOT contain fractional parts.
- */
-export type JsonRpcId = number | string | null;
-
-export interface JsonRpcError {
-  code: number;
-  message: string;
-  data?: unknown;
-  stack?: string;
-}
-
-export interface JsonRpcRequest<T> {
-  jsonrpc: JsonRpcVersion;
-  method: string;
-  id: JsonRpcId;
-  params?: T;
-}
-
-export interface JsonRpcNotification<T> {
-  jsonrpc: JsonRpcVersion;
-  method: string;
-  params?: T;
-}
-
-interface JsonRpcResponseBase {
-  jsonrpc: JsonRpcVersion;
-  id: JsonRpcId;
-}
-
-export interface JsonRpcSuccess<T> extends JsonRpcResponseBase {
-  result: Maybe<T>;
-}
-
-export interface JsonRpcFailure extends JsonRpcResponseBase {
-  error: JsonRpcError;
-}
-
-export type JsonRpcResponse<T> = JsonRpcSuccess<T> | JsonRpcFailure;
-
-export interface PendingJsonRpcResponse<T> extends JsonRpcResponseBase {
-  result?: T;
-  error?: Error | JsonRpcError;
-}
+export type PendingJsonRpcResponse<Result> = Omit<
```

Comment:
Not suggesting any changes, but I wonder if this is type is right, considering that it seems that such a response can theoretically have both `result` and `error`.

Reply:
The type is meant to be used inside middleware functions and internally in `JsonRpcEngine` during request processing. It permits both, either, or neither property being present. Or did you mean something else?